tcp_test.go

```diff
@@ -40,30 +41,125 @@ func (t testTun) WritePacket(p *tuntap.Packet) error {
 	return t.write_error
 }
 
-type testTCPData struct {
+type testNode struct {
 	in          chan types.Packet
 	out         chan types.Packet
 	write_error error
 }
 
-func (d testTCPData) SendPacket(p types.Packet) error {
+func (n testNode) SendPacket(p types.Packet) error {
+	n.in <- p
+	return n.write_error
+}
+func (n testNode) Packets() <-chan types.Packet {
+	return n.out
+}
+func (n testNode) GetNodeAddress() types.NodeAddress {
+	return "source"
+}
+
+func readTunPacket() tuntap.Packet {
+	p := make([]byte, 192, 192)
+	p[0] = 0x6
+	addr := []byte(net.ParseIP("2002::"))
+	copy(p[64:192], addr)
+	tp := tuntap.Packet{Protocol: 0x8000, Truncated: false, Packet: p}
+	return tp
+}
+
+func TestTCPTunRequest(t *testing.T) {
+	if !isRoot() {
+		t.Skip()
+	}
+
+	amt := int64(7)
+	dest := types.NodeAddress("destination")
+	tun := testTun{make(chan *tuntap.Packet), make(chan *tuntap.Packet), nil, nil}
+	node := testNode{make(chan types.Packet), make(chan types.Packet), nil}
+
+	i, err := tuntap.Open("tun%d", tuntap.DevTun)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tcp := NewTCPTunnel(node, tun, dest, amt, i.Name())
+	defer tcp.Close()
+
+	// Check request from new tunnel
+	p := <-node.in
 	var req types.TCPTunnelRequest
-	err := req.UnmarshalBinary(p.Data)
+	err = req.UnmarshalBinary(p.Data)
 	if err != nil {
-		resp := types.TCPTunnelResponse{net.ParseIP("fe80::")}
-		resp_b, _ := resp.MarshalBinary()
-		d.out <- types.Packet{"test", 7, resp_b}
-		return nil
-	} else {
-		d.in <- p
-		return d.write_error
+		t.Fatal(err)
+	}
+
+	select {
+	case err := <-tcp.Error():
+		t.Fatal(err)
+	default:
 	}
 }
-func (d testTCPData) Packets() <-chan types.Packet {
-	return d.out
+
+func TestTCPTunResponse(t *testing.T) {
+	if !isRoot() {
+		t.Skip()
+	}
+
+	amt := int64(7)
+	dest := types.NodeAddress("destination")
+	tun := testTun{make(chan *tuntap.Packet), make(chan *tuntap.Packet), nil, nil}
+	node := testNode{make(chan types.Packet), make(chan types.Packet), nil}
+
+	i, err := tuntap.Open("tun%d", tuntap.DevTun)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Start new client
+	tcp := NewTCPTunnel(node, tun, dest, amt, i.Name())
+	defer tcp.Close()
+
+	// Check request from new client
+	p := <-node.in
+	var req types.TCPTunnelRequest
+	err = req.UnmarshalBinary(p.Data)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Have server send back response
+	resp := types.TCPTunnelResponse{net.ParseIP("2001::")}
+	resp_b, _ := resp.MarshalBinary()
+	p = types.Packet{"source", amt, resp_b}
+	node.out <- p
+
+	time.Sleep(300 * time.Millisecond)
+
+	// Make sure the device is set up with the correct address
+	tun_name := strings.Trim(i.Name(), "\x00")
+
+	ifi, err := net.InterfaceByName(tun_name)
+	if err != nil {
+		t.Fatal(err)
+	}
```

Comment:
I don't think this will work without root permissions. Can you just DI the class which does this?